### PR TITLE
Issue 2052: (SegmentStore) HDFSStorage fencing with runaway instance

### LINKED ISSUE #2052: HDFSStorage potential corruption due to zombie processes
**Problem description**
Suppose we have two instances of the `HDFSStorage` for the same Container, one with epoch A and one with epoch B = A+1, each running on different hosts. Call them HA and HB. HA and HB should never run concurrently if we do a graceful container failover, however it is possible that they are both running at the same time, and writing to the same Segments, if HA's host loses connection to ZK/BK and still thinks it's the leader for the Container. In that case, both HA and HB will be competing to modify the same Segment, thus leading to issues. 

Most of these issues are resolved using the fencing mechanism built into `HDFSStorage`. However, there is a case that is not handled:
1. HA has an ongoing write to Segment X.
2. While the write is happening, HB open-writes Segment X, which causes the last file of X to be marked as read-only, and after that, its length is retrieved.
3. HA completes the write to Segment X, altering it (and increasing its length) after it was marked as read-only.
4. HB now has the wrong information about the state of Segment X, because one of the files has been modified behind the scenes without it knowing about.

This can lead to a number of issues. See **Example Problematic Stack Traces** below for observed issues.

**Problem location**
HDFSStorage.java

**Suggestions for an improvement**
This is due to the fact that setting permissions in HDFS is not atomic, or even serializable with other writes. They can happen concurrently and any permission change will not affect ongoing writes.

We need to make `FileSystemOperation.makeReadOnly` provide a guarantee that, after its successful completion, the target file will never change. This means it should be able to wait for ongoing writes, if possible. Note that due to the fact that there is no known server-side feature to support this, any solution may not work well in a general case, but it should solve the use case that the Segment Store is concerned about (on each Container there is a single StorageWriter accessing a Segment, and we can only solve this for two SegmentContainer zombie instances competing for the same Segment).
1. We introduce a new XAttr named "WritingInProgress"
2. During a write, we set this XAttr (to `true`) on the target file right before we are about to open the write stream, and clear it when we're done (whether successfully or not).
3. `makeReadOnly` will check for this XAttr's existence after it marks the file as read-only. If it's set, it will need to wait until it's cleared (or a timeout expires).

This should make`makeReadOnly` provide enough guarantees that the file will not change after it returns successfully and the rest of the HDFS fencing code in the SegmentStore should work as expected.

**Example Problematic Stack Traces**
```
java.util.concurrent.CompletionException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.HadoopIllegalArgumentException): concat: source file /Segment_1000_3 is invalid or empty or underConstruction
	at org.apache.hadoop.hdfs.server.namenode.FSDirConcatOp.verifySrcFiles(FSDirConcatOp.java:160)
	at org.apache.hadoop.hdfs.server.namenode.FSDirConcatOp.concat(FSDirConcatOp.java:68)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.concat(FSNamesystem.java:1855)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.concat(NameNodeRpcServer.java:983)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.concat(ClientNamenodeProtocolServerSideTranslatorPB.java:579)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorageTest.testZombies(HDFSStorageTest.java:105)
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.HadoopIllegalArgumentException): concat: source file /Segment_1000_3 is invalid or empty or underConstruction
	at org.apache.hadoop.hdfs.server.namenode.FSDirConcatOp.verifySrcFiles(FSDirConcatOp.java:160)
	at org.apache.hadoop.hdfs.server.namenode.FSDirConcatOp.concat(FSDirConcatOp.java:68)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.concat(FSNamesystem.java:1855)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.concat(NameNodeRpcServer.java:983)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.concat(ClientNamenodeProtocolServerSideTranslatorPB.java:579)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.hdfs.DistributedFileSystem.concat(DistributedFileSystem.java:676)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorageTest$FileSystemFixer.concat(HDFSStorageTest.java:217)
	at io.pravega.segmentstore.storage.impl.hdfs.FileSystemOperation.combine(FileSystemOperation.java:196)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.cleanup(OpenWriteOperation.java:140)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.fenceOut(OpenWriteOperation.java:123)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:68)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:24)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorage.lambda$supplyAsync$1(HDFSStorage.java:233)


java.util.concurrent.CompletionException: io.pravega.segmentstore.storage.impl.hdfs.SegmentFilesCorruptedException: Segment 'Segment' has invalid file 'hdfs://localhost:36964/Segment_1000_4'. Declared offset is '1000' but should be '2000'.
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:375)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1934)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorageTest.testZombies(HDFSStorageTest.java:110)
Caused by: io.pravega.segmentstore.storage.impl.hdfs.SegmentFilesCorruptedException: Segment 'Segment' has invalid file 'hdfs://localhost:36964/Segment_1000_4'. Declared offset is '1000' but should be '2000'.
	at io.pravega.segmentstore.storage.impl.hdfs.FileSystemOperation.findAll(FileSystemOperation.java:113)
	at io.pravega.segmentstore.storage.impl.hdfs.FileSystemOperation.checkForFenceOut(FileSystemOperation.java:133)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.fenceOut(OpenWriteOperation.java:103)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:68)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:24)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorage.lambda$supplyAsync$1(HDFSStorage.java:233)


java.util.concurrent.CompletionException: java.lang.AssertionError: New file list would not be contiguous. Combined file ends at offset 3000; Last file begins at 2000.
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:375)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1934)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorageTest.testZombies(HDFSStorageTest.java:110)
Caused by: java.lang.AssertionError: New file list would not be contiguous. Combined file ends at offset 3000; Last file begins at 2000.
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.cleanup(OpenWriteOperation.java:143)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.fenceOut(OpenWriteOperation.java:123)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:68)
	at io.pravega.segmentstore.storage.impl.hdfs.OpenWriteOperation.call(OpenWriteOperation.java:24)
	at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorage.lambda$supplyAsync$1(HDFSStorage.java:233)
```

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
Fixed a bug in `HDFSStorage` where it would not correctly fence out a runaway instance which doesn't know that has been fenced out (aka, zombie instance). 

The HDFS fencing algorithm relies on a basic, yet crucial condition: when we mark a file as read-only, upon return of the method, that file can no longer change. However, with HDFS this is not the case if we have an ongoing (concurrent) write to that file. As an example:
1. A write of length N is initiated at offset L in file F.
2. Shortly thereafter, file F's permissions are updated to make it read-only. This method returns, and `HDFSStorage` may do other operations, such as getting the file's current length (L).
3. The write we previously initiated finishes, and file F now has a length L+N. As such, the file was modified behind the scenes, and the information obtained in the previous step is now obsolete.

This behavior is due to the nature of HDFS, and the fact that a write first goes to the NameNode to get metadata about the file, then sends the data to the Data Node, while a permission change only updates the NameNode. If the write is already in the process of being transmitted to the DataNode, then the permission change will not affect it and it will complete successfully. For any new writes issued after the permissions were set, the behavior is as expected: they will be rejected.

Due to these limitations, we propose the following changes to HDFSStorage and its underlying classes:
- Introduce a new user attribute (`user.writing`).
- In `WriteOperation.java`, 
    - Just prior to calling `FileSystem.append` (and getting an `OutputStream`), set this attribute on the file. This indicates we are about to make a write at one point in the future.
    - Immediately after closing the `OutputStream` opened above, or in case of failure, clear this attribute on the file. This indicates the write completed at one point in the past, whether successfully or not
- In `FileSystemOperation.java`, method `makeReadOnly` (this is invoked by `OpenWrite` - the fencing method)
    - Immediately after updating the permissions on the file, check if the `user.writing` attribute is set on the file. If set, wait as long as it's set, or a maximum amount of time.
        - We need to have a max amount of time because the process writing to the file may have died prior to clearing it, at which point the attribute will never be cleared.
- The above two changes will guarantee that once `makeReadOnly` returns, the file will not change anymore and thus its info can be safely relied upon for fencing purposes.
    - The only exception to this case is if a write takes abnormally long to process (more than 30seconds or whatever the timeout will be). However, this is a very unusual scenario and if it happens, it's likely that the entire system is experiencing slowdowns and all operations will be affected.

**Other Changes**
Added a new type of retry to `Retry` that retries based on a condition and not based on an exception.

**Purpose of the change**
Fixes #2052.

**What the code does**
See **Change log description**

**How to verify it**
Added a new unit test in `HDFSStorageTest` that verifies this zombie scenario.

### CODE REVIEW

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/WriteOperation.java [hunk 1]
```diff
@@ -54,35 +54,41 @@ public void run() throws BadOffsetException, IOException, StorageNotPrimaryExcep
...
-                throw new BadOffsetException(handle.getSegmentName(), lastFile.getLastOffset(), this.offset);
-            } else if (stream.getPos() != lastFile.getLength()) {
-                // Looks like the filesystem changed from underneath us. This could be our bug, but it could be something else.
-                // Update our knowledge of the filesystem and throw a BadOffsetException - this should cause upstream code
-                // to try to reconcile; if it can't then the upstream code should shut down or take other appropriate measures.
-                log.warn("File changed detected for '{}'. Expected length = {}, actual length = {}.", lastFile, lastFile.getLength(), stream.getPos());
-                lastFile.setLength(stream.getPos());
-                throw new BadOffsetException(handle.getSegmentName(), lastFile.getLastOffset(), this.offset);
-            }
+        try {
```
Comment (tkaitchuck) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/WriteOperation.java]: This method is getting long and deep. Perhaps a private helper method is in order.

**Comment by fpj** [comment]
@andreipaduroiu The problem I have with the approach taken in this PR is due to it not matching my understanding of how HDFS works, so perhaps you can shed some light to help me understand why it is a good idea to go down this path.

Let me quote your example:


> 1. A write of length N is initiated at offset L in file F.
> 2. Shortly thereafter, file F's permissions are updated to make it read-only. This method returns, and HDFSStorage may do other operations, such as getting the file's current length (L).
> 3. The write we previously initiated finishes, and file F now has a length L+N. As such, the file was modified behind the scenes, and the information obtained in the previous step is now obsolete.

I'm not sure why such a case can happen because of the following reasoning:

>  1. A write of length N is initiated at offset L in file F.

Opening a file for writing can only succeed in the case the namenode can grant to the client the permission to write to the file. If the open call succeeds, then the client is the exclusive writer.

> 2. Shortly thereafter, file F's permissions are updated to make it read-only. This method returns, and HDFSStorage may do other operations, such as getting the file's current length (L).

This is a metadata operation, so it only touches the namenode. Frankly, I'm not sure whether the namenode allows a change of permission while the lease belongs to a different client. Assuming that the permission update happens, it is very possible that the concurrent write is still going. The datanodes do not know of what is going in the namenode.

 > 3. The write we previously initiated finishes, and file F now has a length L+N. As such, the file was modified behind the scenes, and the information obtained in the previous step is now obsolete.

As I understand it, the write of the block(s) the client is trying to write cannot complete if it doesn't close, which is an operation that touches the namenode. I see two possible scenarios here:

a. The concurrent writer fails in Step 3 when trying to complete the write because the file has been made read only.
b. The concurrent writer crashes before it tries to complete the write, in which case, the blocks written to the datanodes will never really be part of the file.

**Comment by andreipaduroiu** [comment]
> Opening a file for writing can only succeed in the case the namenode can grant to the client the permission to write to the file. If the open call succeeds, then the client is the exclusive writer.

**Comment by andreipaduroiu** [comment]
This was presumably addressed with #2401

**Pull request closed without merging**